### ISSUE TITLE
✍️ Contribution period: Sohier

### ISSUE THREAD

**Issue by soh-123** [ersilia-os/ersilia#875]
### Week 1 - Get to know the community

- [X] Join the communication channels
- [X] Open a GitHub issue (this one!)
- [x] Install the Ersilia Model Hub and test the simplest model
- [x] Write a motivation statement to work at Ersilia
- [x] Submit your first contribution to the Outreachy site

### Week 2 - Install and run an ML model

- [x] Select a model from the suggested list
- [x] Install the model in your system
- [x] Run predictions for the EML
- [x] Compare results with the Ersilia Model Hub implementation!
- [x] Install and run Docker!

### Week 3 - Propose new models

- [x] Suggest a new model and document it (1)
- [x] Suggest a new model and document it (2)
- [x] Suggest a new model and document it (3)

### Week 4 - Prepare your final application

- [x] Submit the final application in the Outreachy website

**Comment by soh-123** [comment]
### Motivation letter

Greetings to all members of the Ersilia Model Hub community,

My name is Sohier and I am a junior software engineer who has recently shifted from UX design. I am writing to express my keen interest in joining the Ersilia Model Hub as an intern software engineer.

After reading through the project description, I am deeply inspired by the mission of providing an open-source, low-code solution dedicated to infectious and neglected disease research, and the use of AI/ML models for drug discovery. The unique blend of technology and healthcare, both areas I am deeply interested in, has drawn me to this incredible opportunity.
I am impressed by the fact that Ersilia Model Hub not only integrates published models from literature but also encourages the development of custom models from its community. This not only ensures the diversity of the models but also fosters a strong sense of community among the contributors. As an aspiring software engineer, I am eager to contribute to both aspects of model integration, and my academic background coupled with my enthusiasm to learn will make me a valuable addition to your team.

I am currently a self-taught software engineer, eager to explore different areas in the field and gain expertise in them. I am keen on applying my skills in Python, Bash, and Conda to work on a fascinating project like Ersilia. I believe I have the technical proficiency required to contribute to the project immediately. The idea of utilizing my skills for a noble cause, such as eradicating infectious diseases, fills me with excitement and motivation.

If selected, I am confident in my ability to:
- Collaborate with the existing team to enhance the platform's functionality and performance.
- Help in addressing and resolving any open issues, as well as identify potential areas of improvement.
- Engage with the community, understanding their needs, and making the platform more user-friendly and accessible.
- 
I strongly believe that the Ersilia Open Source Initiative is a ray of hope for labs, universities, and clinics, particularly in low- and middle-income countries. I am enthusiastic about contributing to this mission and utilizing my software engineering abilities to make a real impact.

Thank you for considering my application. I look forward to being part of this community.

**Comment by soh-123** [comment]
### Plasma Protein Binding (IDL-PPBopt)

As far as I know, it's studying the improvement of drug efficiency by predicting and optimizing how compounds bind to human plasma proteins, Deep learning can potentially enhance a drug's therapeutic effects and reduce side effects. This can lead to more effective and safer medications for patients. I found that it would be a potential model to work on.

**Comment by soh-123** [comment]
### Update

Installing this project had some issues, I found out I needed to install torch using Cuda, which is available for specific some devices and mine was not one of them (MacBook Air) it's listed [on their website](https://developer.nvidia.com/nvidia-cuda-toolkit-11_7_0-developer-tools-mac-hosts#:~:text=NVIDIA%C2%AE%20CUDA%20Toolkit%2011.7,sessions%20on%20supported%20target%20platforms.)

because I'm in a tight time, I decided to move on to another project and read the install instructions thoroughly before choosing it.

**Comment by soh-123** [comment]
## Task 6: Installing STOUT model :

The steps were straightforward, and I got some experience from my trials in the previous project.

I created Stout environment:
`conda create --name STOUT python=3.8 
`
then activate it 
`conda activate STOUT
`
Then install the model itself
`conda install -c decimer stout-pypi
`
Finally, test
`
from STOUT import translate_forward, translate_reverse

# SMILES to IUPAC name translation

SMILES = "CN1C=NC2=C1C(=O)N(C(=O)N2C)C"
IUPAC_name = translate_forward(SMILES)
print("IUPAC name of "+SMILES+" is: "+IUPAC_name)

# IUPAC name to SMILES translation

IUPAC_name = "1,3,7-trimethylpurine-2,6-dione"
SMILES = translate_reverse(IUPAC_name)
print("SMILES of "+IUPAC_name+" is: "+SMILES)
`

**Comment by soh-123** [comment]
### Task 7: Run Prediction from EML

I spent sometimes, figuring out where should I start, and I found it's simply importing STOUT in Python, reading the EML, and creating the output file. 
I liked to split steps for clarity, so I took the eml file as input and extracted the smiles then translated it to iupac output

```python
import pandas as pd

# Read the EML data
eml_data = pd.read_csv("eml.csv")

# Extract and save SMILES strings
smiles_data = eml_data["smiles"]
smiles_data.to_csv("smiles.csv", index=False)

from STOUT import translate_forward

# Read SMILES data
with open("smiles.csv", "r") as f:
    smiles_list = f.readlines()[1:]  # skip the header

iupac_names = [translate_forward(smile.strip()) for smile in smiles_list]

# Save IUPAC names
with open("iupac.csv", "w") as f:
    for name in iupac_names:
        f.write(name + "\n")
``` 

The code output:
The output of the code is a CSV file of IUPAC names

**Comment by soh-123** [comment]
### Task 8: understand Ersilia's backend and installing Docker

I already have docker installed on my device so I only set ersilia on docker:
- navigate to the directory where 'dockerfile' is.
- build a docker image 
`
docker build -t ersilia-image .
`
- run the docker container.
- I tried to fetch and then serve different models, each time you serve a model you are creating a container on docker.
- finally, I tried to fetch and serve STOUT, but it didn't work and showed me this error message:
`
docker.errors.ImageNotFound: 404 Client Error for http+docker://localhost/v1.43/images/ersiliaos/eos4se9:latest/json: Not Found ("No such image: ersiliaos/eos4se9:latest")
`
I'll keep trying tomorrow and ask about this error in slack ..

**Comment by soh-123** [comment]
### Compare results with the Ersilia models

In this task, I'll run the model in Ersilia hub and compare the prediction data from task 7 to this output

Run prediction
`ersilia api run -i eml.csv -o ersilia_iupac_prediction.csv`

after waiting a few hours I get the output file : `ersilia_iupac_prediction.csv`

Here's a snippet of it:
`
key,input,iupacs_names
MCGSCOLBFJQGHM-SCZZXKLOSA-N,Nc1nc(NC2CC2)c3ncn([C@@H]4C[C@H](CO)C=C4)c3n1,"[(1R,4R)-4-[2-amino-4-(cyclopropylamino)-4H-purin-9-yl>
GZOSMCIZMLWJML-VJLLXTKPSA-N,C[C@]12CC[C@H](O)CC1=CC[C@@H]3[C@@H]2CC[C@@]4(C)[C@H]3CC=C4c5cccnc5,"(1S,2S,5S,10R,11R,14S)-5,11-dime>
BZKPWHYZMXOIDC-UHFFFAOYSA-N,CC(=O)Nc1sc(nn1)[S](N)(=O)=O,"N-[5-[amino(dioxo)-λ6-thia-3,4-diazacyclopent-2-en-2-yl]acetamide"
QTBSBXVTEAMEQO-UHFFFAOYSA-N,CC(O)=O,aceticacid
PWKSKIMOESPYIA-BYPYZUCNSA-N,CC(=O)N[C@@H](CS)C(O)=O,(2R)-2-acetamido-3-sulfanylpropanoicacid
BSYNRYMUTXBXSQ-UHFFFAOYSA-N,CC(=O)Oc1ccccc1C(O)=O,2-acetyloxybenzoicacid
MKUXAQIIEYXACX-UHFFFAOYSA-N,NC1=NC(=O)c2ncn(COCCO)c2N1,2-amino-9-(2-hydroxyethoxymethyl)-3H-purin-6-one
`

I took the output file from task 7: `iupac.csv`

**Comment by soh-123** [comment]
# 1st Model Suggestion

### GlyLES: Grammar-based Parsing of Glycans from IUPAC-condensed to SMILES

Publication: [PubMed](https://pubmed.ncbi.nlm.nih.gov/36959676/)

Published: March 2023

Authors: Roman Joeres, Daniel Bojar, Olga V Kalinina

Source Code: [Github-GlyLES](https://github.com/kalininalab/GlyLES/)

### About The Model

Glycans are complex sugar molecules found on the surfaces of cells. They are often attached to proteins (glycoproteins) and lipids (glycolipids) and play important roles in various cellular functions. For example, they are involved in protein folding, interactions between cells, and other processes that occur outside of cells. Changes in the structure of glycans can have implications for diseases like infections and cancer.
GlyLES is a tool to convert IUPACs into smile representations of the Glycans, the limitations of the other tools that the authors found were the accuracy of the interpretation of Glycans and the visualization feature. This conversion is important for various applications in glycobiology, such as visualizing glycan structures, searching for specific substructures within glycans, performing molecular modeling and docking studies, and developing machine learning algorithms.

### Why does it matter to Ersilia
Glycans play an important role in dealing with cancer and infections, it's important for scientists to see accurate data on glycans and a visual representation would make it easier and time efficient.

**Comment by soh-123** [comment]
# 2nd Model Suggestion

### Inferring the sources of HIV infection in Africa from deep sequence data with semi-parametric Bayesian Poisson flow models

**Publication**: [arXiv community](https://arxiv.org/abs/2110.12273v4)

**Published**: January 2022

**Authors**: Xiaoyue Xi, Simon EF Spencer, Matthew Hall, M Kate Grabowski, Joseph Kagaayi, Oliver Ratmann

**Source Code**: [Github-phyloscanner](https://github.com/bdi-pathogens/phyloscanner)

### About The Model

Researchers are using a powerful technology called "pathogen deep-sequencing" to study HIV in African populations. This technology allows them to examine the genetic makeup of the virus in great detail. They have developed a complex mathematical model (called a "semi-parametric Bayesian Poisson model") to analyze this genetic data. This model helps them figure out how the virus is transmitted between people and where individuals might have gotten infected.
One of the key advantages of this model is that it can handle a lot of data and provide detailed insights. It also takes into account factors like potential biases in the data and can estimate how HIV spreads among different groups of people, such as men and women of different ages.

### Why does it matter to Ersilia

This research studies the genetic data of HIV diseases, collects the pathogens fragments of the infection as puzzle pieces for different hosts, and creates family trees to understand how these pathogens evolve and spread.

**Comment by soh-123** [comment]
# 3rd Model Suggestion

### IAtomic Convolutional Networks for Predicting Protein-Ligand Binding Affinity

**Publication**: [arXiv community](https://arxiv.org/abs/1703.10603v1)

**Published**: March 2017

**Authors**: Joseph Gomes, Bharath Ramsundar, Evan N. Feinberg, Vijay S. Pande

**Source Code**: [Github-DeepChem](https://github.com/deepchem/deepchem)

### About The Model

This model focused on predicting how well a drug-like molecule binds to a specific target (like a protein) in the body. When scientists are developing new drugs, they need to know how well a molecule will attach itself to a specific target in the body. This attachment is crucial for the drug to work effectively.
The problem is that the current methods are "Molecular force fields which are complex mathematical equations" and 
"Cheminformatics descriptors which are characteristics of molecules that are hand-crafted by experts", these methods rely on human knowledge and can be limited. Scientists have to manually select which features to use, and this can be challenging.
In this research, there's a new method called "atomic convolutional neural network." It's a type of computer program that can automatically learn how molecules interact with each other by looking at their 3D structures.
The program is trained using data on the binding strength of molecules to proteins. It calculates the energy associated with the interaction between molecules, proteins, and ligands (drug-like molecules) using their atomic structures.

### Why does it matter to Ersilia

This model is meant to test the effect of the drug on proteins in the human body and the trained data will be helpful to be integrated with other modules.

**Comment by GemmaTuron** [comment]
Hello,

Thanks for your work during the Outreachy contribution period, we hope you enjoyed it! We will now close this issue while we work on the selection of interns. Thanks again!
